src/lib/Simplify.hs

```diff
@@ -30,183 +27,204 @@ data SimpOpts = SimpOpts { preserveDerivRules :: Bool }
 -- variables have been eliminated. In principle the substitution with global env is optional, but
 -- in practice it also has to happen sooner or later, because the Imp pass assumes that only array
 -- variables are imported.
-type SimplifyM a = ReaderT SimpEnv (ReaderT ((SubstEnv, RuleEnv), SimpOpts) Embed) a
+type SimplifyM a = ReaderT SimpEnv (ReaderT (SubstEnv, SimpOpts) Embed) a
 
-simplifyModule :: SubstEnv -> RuleEnv -> Module -> (Module, SubstEnv)
-simplifyModule substEnv rulesEnv m = (mOut', subst (envOut', mempty) envOut)
-  where (mOut , envOut ) = simplifyModuleOpts (SimpOpts True ) (substEnv, rulesEnv) m
-        (mOut', envOut') = simplifyModuleOpts (SimpOpts False) (substEnv, rulesEnv) mOut
+simplifyModule :: SubstEnv -> Module -> (Module, SubstEnv)
```

Comment:
Do any simplification contracts change after this PR? IIRC you said that your change will make it optional to inline the atoms instead of the previous behavior which has forced it. I don't remember what it was exactly, but does this change anything?

Reply:
Yes, it's now syntactically possible to have atoms on the rhs of a decl, but it shouldn't happen after this pass. As we discussed, I'll write a checker for the simplified IR that documents and checks these rules.